File: pascalsTriangle.cpp
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> generate(int numRows) {
        std::vector<std::vector<int>> triangle;
        std::vector<int> prev, curr;

        for (int i = 0; i < numRows; ++i) {
            curr.clear();
            curr.push_back(1); // First element is always 1

            for (int j = 1; j < i; ++j) {
                curr.push_back(prev[j - 1] + prev[j]);
            }

            if (i > 0) {
                curr.push_back(1); // Last element is always 1
            }

            triangle.push_back(curr);
            std::swap(prev, curr);
        }

        return triangle;
    }
};

File: pascalsTriangle.rs
impl Solution {
    pub fn generate(num_rows: i32) -> Vec<Vec<i32>> {
        let num_rows = num_rows as usize;
        let mut triangle: Vec<Vec<i32>> = Vec::with_capacity(num_rows);

        // Buffers for current and previous rows
        let mut prev: Vec<i32> = Vec::new();
        let mut curr: Vec<i32> = Vec::new();

        for i in 0..num_rows {
            curr.clear();
            curr.push(1); // First element is always 1

            for j in 1..i {
                curr.push(prev[j - 1] + prev[j]);
            }

            if i > 0 {
                curr.push(1); // Last element is always 1
            }

            // Push a clone of the current row
            triangle.push(curr.clone());

            // Swap buffers for next iteration
            std::mem::swap(&mut prev, &mut curr);
        }

        triangle
    }
}
